node/lib/DeviceAccessor.ts

```diff
@@ -0,0 +1,241 @@
+
+import { DeviceInterface } from "./DeviceInterface";
+import { IDevice } from "./ITranslator";
+
+import { EventEmitter } from "events";
+
+/**
+ * Provides reflection-style access to device properties and methods via device interfaces.
+ */
+export class DeviceAccessor {
+
+    /**
+     * Gets the value of a property on a device, using a specified interface.
+     *
+     * @param {IDevice} device  Device instance obtained from a translator
+     * @param {(string | DeviceInterface)} interfaceName  Name of the interface used
+     *     to access the device, or a DeviceInterface object
+     * @param {string} propertyName  Name of the property to get
+     * @returns {Promise<any>} Value returned by the property getter; the value is
+     *     expected to conform to the JSON schema for the property as specified
+     *     by the interface
+     */
+    public static async getPropertyAsync(
+            device: IDevice,
+            interfaceName: string | DeviceInterface,
+            propertyName: string): Promise<any> {
+        let deviceInterface = DeviceAccessor.getDeviceInterface(device, interfaceName);
+        DeviceAccessor.validateMemberName(propertyName);
+
+        let value: any = deviceInterface[propertyName];
+        if (typeof value === "undefined") {
+            let methodName = "get" + DeviceAccessor.capitalize(propertyName);
+            let getPropertyMethod: any = deviceInterface[methodName];
+            if (typeof getPropertyMethod === "function") {
+                value = getPropertyMethod.call(deviceInterface);
+            } else {
+                getPropertyMethod = deviceInterface[methodName + "Async"];
+                if (typeof getPropertyMethod === "function") {
+                    value = getPropertyMethod.call(deviceInterface);
```

Comment:
The design allows for device methods to be optionally async, so they could return a value directly instead of a Promise. The code on lines 47-48 below does the await only if appropriate.

---
In reply to: [71902252](https://github.com/opent2t/opent2t/pull/1#discussion_r71902252) [](ancestors = 71902252)